File: solutions/typescript/matrix/3/matrix.ts
export class Matrix {
  #rows: number[][];

  constructor(inputText: string) {
    this.#rows = inputText.split('\n').map((inputRow: string) =>
      inputRow.split(' ').map((token: string) => +token)
    )
  }

  get rows(): number[][] {
    return this.#rows
  }

  get columns(): number[][] {
    return this.#rows[0].map((_, c: number) => 
      this.#rows.map((row: number[]) => row[c])
    );
  }
}

File: solutions/prolog/matrix/1/matrix.pl
string_number(String, Number) :- number_string(Number, String).

string_numlist(String, NumList) :-
    split_string(String, " ", "", StringList),
    maplist(string_number, StringList, NumList).

rows(Matrix, Rows) :-
    split_string(Matrix, "\n", "", RawRows),
    maplist(string_numlist, RawRows, Rows).

row(Matrix, Y, Row) :-
    rows(Matrix, Rows),
    nth1(Y, Rows, Row).

column(Matrix, X, Column) :-
    rows(Matrix, Rows),
    maplist(nth1(X), Rows, Column).
